warehouse/query-core/src/main/java/datawave/query/postprocessing/tf/DocumentKeysFunction.java

```diff
@@ -0,0 +1,220 @@
+import java.util.HashSet;
+import java.util.Set;
+
+/**
+ * A function that intersects document keys prior to fetching term offsets.
+ * <p>
+ * This only works on positive phrase queries -- that is, no guarantee is made for negated phrases
+ */
+public class DocumentKeysFunction {
+    
+    private static final Logger log = Logger.getLogger(DocumentKeysFunction.class);
+    
+    // cannot intersect on negated values
+    private final Set<String> negatedValues = new HashSet<>();
+    private final Multimap<String,ContentFunction> contentFunctions = LinkedListMultimap.create();
+    
+    /**
+     * @param config
+     *            a {@link TermFrequencyConfig}
+     */
+    public DocumentKeysFunction(TermFrequencyConfig config) {
+        populateContentFunctions(config.getScript());
+    }
+    
+    protected void populateContentFunctions(JexlNode node) {
+        
+        ContentFunctionsDescriptor descriptor = new ContentFunctionsDescriptor();
+        ContentJexlArgumentDescriptor argsDescriptor;
+        Set<String>[] fieldsAndTerms;
+        JexlNode parent;
+        
+        Multimap<String,Function> functions = TermOffsetPopulator.getContentFunctions(node);
+        for (String key : functions.keySet()) {
+            Collection<Function> coll = functions.get(key);
+            for (Function f : coll) {
+                parent = f.args().get(0).jjtGetParent();
+                
+                if (!(parent instanceof ASTFunctionNode)) {
+                    throw new IllegalArgumentException("parent was not a function node");
+                }
+                
+                argsDescriptor = descriptor.getArgumentDescriptor((ASTFunctionNode) parent);
+                
+                // content, tf, and indexed fields are not actually needed to extract fields from the function node
+                fieldsAndTerms = argsDescriptor.fieldsAndTerms(Collections.emptySet(), Collections.emptySet(), Collections.emptySet(), null);
+                
+                if (fieldsAndTerms[0].size() != 1) {
+                    throw new IllegalStateException("content function had more than one field");
+                }
+                
+                ContentFunction contentFunction = new ContentFunction(fieldsAndTerms[0].iterator().next(), fieldsAndTerms[1]);
+                contentFunctions.put(contentFunction.getField(), contentFunction);
+                
+                // need to record any values that are part of a negative function
+                if (isFunctionNegated(f)) {
+                    negatedValues.addAll(contentFunction.getValues());
+                }
+            }
+        }
+    }
+    
+    private boolean isFunctionNegated(Function f) {
+        JexlNode node = f.args().get(0);
+        if (node != null) {
+            while (node.jjtGetParent() != null) {
+                node = node.jjtGetParent();
+                if (node instanceof ASTNotNode) {
+                    return true;
```

Comment:
Might we handle a not within a not?  This should travel all the way to the top of the tree.

Reply:
JexlASTHelper.isNegated(JexlNode node) might be in order.  I am sure there are many places we could be using that.

Reply:
Just documenting our earlier conversation: we shouldn't need special handling of negations in this case.